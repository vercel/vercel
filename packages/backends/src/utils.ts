import { delimiter } from 'path';
import { dirname, join } from 'path';
import {
  download,
  runNpmInstall,
  runPackageJsonScript,
  getNodeVersion,
  getSpawnOptions,
  execCommand,
  getEnvForPackageManager,
  scanParentDirs,
  getNodeBinPaths,
} from '@vercel/build-utils';
import type { BuildV2 } from '@vercel/build-utils';

export async function downloadInstallAndBundle(args: Parameters<BuildV2>[0]) {
  const { entrypoint, files, workPath, meta, config, repoRootPath } = args;
  await download(files, workPath, meta);

  const entrypointFsDirname = join(workPath, dirname(entrypoint));
  const nodeVersion = await getNodeVersion(
    entrypointFsDirname,
    undefined,
    config,
    meta
  );

  const spawnOpts = getSpawnOptions(meta || {}, nodeVersion);

  const {
    cliType,
    lockfileVersion,
    packageJsonPackageManager,
    turboSupportsCorepackHome,
  } = await scanParentDirs(entrypointFsDirname, true, repoRootPath);

  spawnOpts.env = getEnvForPackageManager({
    cliType,
    lockfileVersion,
    packageJsonPackageManager,
    nodeVersion,
    env: spawnOpts.env || {},
    turboSupportsCorepackHome,
    projectCreatedAt: config.projectSettings?.createdAt,
  });

  const installCommand = config.projectSettings?.installCommand;
  if (typeof installCommand === 'string') {
    if (installCommand.trim()) {
      console.log(`Running "install" command: \`${installCommand}\`...`);
      await execCommand(installCommand, {
        ...spawnOpts,
        cwd: entrypointFsDirname,
      });
    } else {
      console.log(`Skipping "install" command...`);
    }
  } else {
    await runNpmInstall(
      entrypointFsDirname,
      [],
      spawnOpts,
      meta,
      nodeVersion,
      config.projectSettings?.createdAt
    );
  }
  return { entrypointFsDirname, nodeVersion, spawnOpts };
}

export async function maybeExecBuildCommand(
  args: Parameters<BuildV2>[0],
  options: Awaited<ReturnType<typeof downloadInstallAndBundle>>
) {
  const projectBuildCommand = args.config.projectSettings?.buildCommand;
  if (projectBuildCommand) {
    // Add node_modules/.bin to PATH so commands like 'cervel' can be found
    const repoRoot = args.repoRootPath || args.workPath;
    const nodeBinPaths = getNodeBinPaths({
      base: repoRoot,
      start: args.workPath,
    });
    const nodeBinPath = nodeBinPaths.join(delimiter);
    const env = {
      ...options.spawnOpts.env,
      PATH: `${nodeBinPath}${delimiter}${options.spawnOpts.env?.PATH || process.env.PATH}`,
    };

    return execCommand(projectBuildCommand, {
      ...options.spawnOpts,
      env,
      cwd: args.workPath,
    });
  }

  // I don't think we actually want to support vercel-build or now-build because those are hacks for controlling api folder builds
  const possibleScripts = ['build'];

  return runPackageJsonScript(
    options.entrypointFsDirname,
    possibleScripts,
    options.spawnOpts,
    args.config.projectSettings?.createdAt
  );
}
