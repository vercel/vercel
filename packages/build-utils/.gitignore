?.js
dist
test/symlinks-out
test/symlinks.zip
