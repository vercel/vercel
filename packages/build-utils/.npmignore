/src
/test
/tsconfig.json
/.turbo
/jest.config.js
tmp
