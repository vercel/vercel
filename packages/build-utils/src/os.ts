import { readFile } from 'fs-extra';
import { isErrnoException } from '@vercel/error-utils';

export async function getOsRelease() {
  try {
    const data = await readFile('/etc/os-release', 'utf8');
    return await parseOsRelease(data);
  } catch (err) {
    if (isErrnoException(err) && err.code === 'ENOENT') {
      return null;
    }
    throw err;
  }
}

export async function parseOsRelease(data: string) {
  const obj: Record<string, string> = {};
  // Example file contents:
  //   NAME="Amazon Linux"
  //   VERSION="2023"
  //   ID="amzn"
  //   ID_LIKE="fedora"
  for (const line of data.trim().split('\n')) {
    const m = /(?<key>.*)="(?<value>.*)"/.exec(line);
    if (!m?.groups) {
      continue;
    }
    obj[m.groups.key] = m.groups.value;
  }
  return obj;
}

export async function getProvidedRuntime() {
  const os = await getOsRelease();
  if (!os) {
    return 'provided.al2023';
  }

  return os.PRETTY_NAME === 'Amazon Linux 2'
    ? 'provided.al2'
    : 'provided.al2023';
}
