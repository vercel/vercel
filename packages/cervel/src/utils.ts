// Colors support for terminal output
const noColor =
  globalThis.process?.env?.NO_COLOR === '1' ||
  globalThis.process?.env?.TERM === 'dumb';

// map each code to its matching reset code
const resets: Record<number, number> = {
  1: 22,
  31: 39,
  32: 39,
  33: 39,
  34: 39,
  35: 39,
  36: 39,
  90: 39,
};

const _c = (c: number) => (text: string) => {
  if (noColor) return text;
  const off = resets[c] ?? 0;
  return `\u001B[${c}m${text}\u001B[${off}m`;
};

export const Colors = {
  bold: _c(1),
  red: _c(31),
  green: _c(32),
  yellow: _c(33),
  blue: _c(34),
  magenta: _c(35),
  cyan: _c(36),
  gray: _c(90),
  url: (title: string, url: string) =>
    noColor
      ? `[${title}](${url})`
      : `\u001B]8;;${url}\u001B\\${title}\u001B]8;;\u001B\\`,
} as Record<
  | 'bold'
  | 'red'
  | 'green'
  | 'yellow'
  | 'blue'
  | 'magenta'
  | 'cyan'
  | 'gray'
  | 'url',
  (text: string) => string
> & {
  url: (title: string, url: string) => string;
};
