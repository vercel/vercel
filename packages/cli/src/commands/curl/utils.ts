export const requoteArgs = (arg: string): string => {
  if (arg.includes(' ')) {
    if (arg.includes('"')) {
      return `'${arg}'`;
    } else {
      return `"${arg}"`;
    }
  }
  return arg;
};
