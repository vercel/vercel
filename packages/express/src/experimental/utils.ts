import { dirname, join } from 'path';
import {
  download,
  runNpmInstall,
  runPackageJsonScript,
  getNodeVersion,
  getSpawnOptions,
  execCommand,
  getEnvForPackageManager,
  scanParentDirs,
} from '@vercel/build-utils';
import type { BuildV2 } from '@vercel/build-utils';

export async function downloadInstallAndBundle(args: Parameters<BuildV2>[0]) {
  const { entrypoint, files, workPath, meta, config } = args;
  await download(files, workPath, meta);
  const entrypointFsDirname = join(workPath, dirname(entrypoint));
  const nodeVersion = await getNodeVersion(
    entrypointFsDirname,
    undefined,
    config,
    meta
  );
  const spawnOpts = getSpawnOptions(meta || {}, nodeVersion);

  const {
    cliType,
    lockfileVersion,
    packageJsonPackageManager,
    turboSupportsCorepackHome,
  } = await scanParentDirs(entrypointFsDirname, true);

  spawnOpts.env = getEnvForPackageManager({
    cliType,
    lockfileVersion,
    packageJsonPackageManager,
    nodeVersion,
    env: spawnOpts.env || {},
    turboSupportsCorepackHome,
    projectCreatedAt: config.projectSettings?.createdAt,
  });

  const installCommand = config.projectSettings?.installCommand;
  if (typeof installCommand === 'string') {
    if (installCommand.trim()) {
      console.log(`Running "install" command: \`${installCommand}\`...`);
      await execCommand(installCommand, {
        ...spawnOpts,
        cwd: entrypointFsDirname,
      });
    } else {
      console.log(`Skipping "install" command...`);
    }
  } else {
    await runNpmInstall(
      entrypointFsDirname,
      [],
      spawnOpts,
      meta,
      nodeVersion,
      config.projectSettings?.createdAt
    );
  }
  return { entrypointFsDirname, nodeVersion, spawnOpts };
}

export async function maybeExecBuildCommand(
  args: Parameters<BuildV2>[0],
  options: Awaited<ReturnType<typeof downloadInstallAndBundle>>
) {
  const projectBuildCommand = args.config.projectSettings?.buildCommand;
  if (projectBuildCommand) {
    await execCommand(projectBuildCommand, {
      ...options.spawnOpts,
      cwd: args.workPath,
    });
  } else {
    // I don't think we actually want to support vercel-build or now-build because those are hacks for controlling api folder builds
    const possibleScripts = ['build'];

    await runPackageJsonScript(
      options.entrypointFsDirname,
      possibleScripts,
      options.spawnOpts,
      args.config.projectSettings?.createdAt
    );
  }
}
