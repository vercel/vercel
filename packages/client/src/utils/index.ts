import { FilesMap } from './hashes';
import nodeFetch, { RequestInit } from 'node-fetch';
import { join, sep, relative, basename } from 'path';
import { URL } from 'url';
import ignore from 'ignore';
import { pkgVersion } from '../pkg';
import { NowBuildError } from '@vercel/build-utils';
import { VercelClientOptions, VercelConfig } from '../types';
import { Sema } from 'async-sema';
import { readFile } from 'fs-extra';
import readdir from './readdir-recursive';

type Ignore = ReturnType<typeof ignore>;

const semaphore = new Sema(10);

export const API_FILES = '/v2/files';

const EVENTS_ARRAY = [
  // File events
  'hashes-calculated',
  'file-count',
  'file-uploaded',
  'all-files-uploaded',
  // Deployment events
  'created',
  'building',
  'ready',
  'alias-assigned',
  'warning',
  'error',
  'notice',
  'tip',
  'canceled',
  // Checks events
  'checks-registered',
  'checks-completed',
  'checks-running',
  'checks-conclusion-succeeded',
  'checks-conclusion-failed',
  'checks-conclusion-skipped',
  'checks-conclusion-canceled',
] as const;

export type DeploymentEventType = (typeof EVENTS_ARRAY)[number];
export const EVENTS = new Set(EVENTS_ARRAY);

export function getApiDeploymentsUrl() {
  return '/v13/deployments';
}

export async function parseVercelConfig(
  filePath?: string
): Promise<VercelConfig> {
  if (!filePath) {
    return {};
  }

  try {
    const jsonString = await readFile(filePath, 'utf8');

    return JSON.parse(jsonString);
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e);

    return {};
  }
}

const maybeRead = async function <T>(path: string, default_: T) {
  try {
    return await readFile(path, 'utf8');
  } catch (err) {
    return default_;
  }
};

export async function buildFileTree(
  path: string | string[],
  {
    isDirectory,
    prebuilt,
    vercelOutputDir,
  }: Pick<VercelClientOptions, 'isDirectory' | 'prebuilt' | 'vercelOutputDir'>,
  debug: Debug
): Promise<{ fileList: string[]; ignoreList: string[] }> {
  const ignoreList: string[] = [];
  let fileList: string[];
  let { ig, ignores } = await getVercelIgnore(path, prebuilt, vercelOutputDir);

  debug(`Found ${ignores.length} rules in .vercelignore`);
  debug('Building file tree...');

  if (isDirectory && !Array.isArray(path)) {
    // Directory path
    const ignores = (absPath: string) => {
      const rel = relative(path, absPath);
      const ignored = ig.ignores(rel);
      if (ignored) {
        ignoreList.push(rel);
      }
      return ignored;
    };
    fileList = await readdir(path, [ignores]);

    if (prebuilt) {
      // Traverse over the `.vc-config.json` files and include
      // the files referenced by the "filePathMap" properties
      const refs = new Set<string>();
      const vcConfigFilePaths = fileList.filter(
        file => basename(file) === '.vc-config.json'
      );
      await Promise.all(
        vcConfigFilePaths.map(async p => {
          const configJson = await readFile(p, 'utf8');
          const config = JSON.parse(configJson);
          if (!config.filePathMap) return;
          for (const v of Object.values(config.filePathMap) as string[]) {
            refs.add(join(path, v));
          }
        })
      );
      if (refs.size > 0) {
        fileList = fileList.concat(Array.from(refs));
      }
    }

    debug(`Found ${fileList.length} files in the specified directory`);
  } else if (Array.isArray(path)) {
    // Array of file paths
    fileList = path;
    debug(`Assigned ${fileList.length} files provided explicitly`);
  } else {
    // Single file
    fileList = [path];
    debug(`Deploying the provided path as single file`);
  }

  return { fileList, ignoreList };
}

export async function getVercelIgnore(
  cwd: string | string[],
  prebuilt?: boolean,
  vercelOutputDir?: string
): Promise<{ ig: Ignore; ignores: string[] }> {
  const ig = ignore();
  let ignores: string[];

  if (prebuilt) {
    if (typeof vercelOutputDir !== 'string') {
      throw new Error(
        `Missing required \`vercelOutputDir\` parameter when "prebuilt" is true`
      );
    }
    if (typeof cwd !== 'string') {
      throw new Error(`\`cwd\` must be a "string"`);
    }
    const relOutputDir = relative(cwd, vercelOutputDir);
    ignores = ['*'];
    const parts = relOutputDir.split(sep);
    parts.forEach((_, i) => {
      const level = parts.slice(0, i + 1).join('/');
      ignores.push(`!${level}`);
    });
    ignores.push(`!${parts.join('/')}/**`);
    ig.add(ignores.join('\n'));
  } else {
    ignores = [
      '.hg',
      '.git',
      '.gitmodules',
      '.svn',
      '.cache',
      '.next',
      '.now',
      '.vercel',
      '.npmignore',
      '.dockerignore',
      '.gitignore',
      '.*.swp',
      '.DS_Store',
      '.wafpicke-*',
      '.lock-wscript',
      '.env.local',
      '.env.*.local',
      '.venv',
      '.yarn/cache',
      'npm-debug.log',
      'config.gypi',
      'node_modules',
      '__pycache__',
      'venv',
      'CVS',
    ];

    const cwds = Array.isArray(cwd) ? cwd : [cwd];

    const files = await Promise.all(
      cwds.map(async cwd => {
        const [vercelignore, nowignore] = await Promise.all([
          maybeRead(join(cwd, '.vercelignore'), ''),
          maybeRead(join(cwd, '.nowignore'), ''),
        ]);
        if (vercelignore && nowignore) {
          throw new NowBuildError({
            code: 'CONFLICTING_IGNORE_FILES',
            message:
              'Cannot use both a `.vercelignore` and `.nowignore` file. Please delete the `.nowignore` file.',
            link: 'https://vercel.link/combining-old-and-new-config',
          });
        }
        return vercelignore || nowignore;
      })
    );

    const ignoreFile = files.join('\n');

    ig.add(`${ignores.join('\n')}\n${clearRelative(ignoreFile)}`);
  }

  return { ig, ignores };
}

/**
 * Remove leading `./` from the beginning of ignores
 * because ignore doesn't like them :|
 */
function clearRelative(str: string) {
  return str.replace(/(\n|^)\.\//g, '$1');
}

interface FetchOpts extends RequestInit {
  apiUrl?: string;
  method?: string;
  teamId?: string;
  headers?: { [key: string]: any };
  userAgent?: string;
}

export const fetch = async (
  url: string,
  token: string,
  opts: FetchOpts = {},
  debugEnabled?: boolean
): Promise<any> => {
  semaphore.acquire();
  const debug = createDebug(debugEnabled);
  let time: number;

  url = `${opts.apiUrl || 'https://api.vercel.com'}${url}`;
  delete opts.apiUrl;

  const { VERCEL_TEAM_ID } = process.env;

  if (VERCEL_TEAM_ID) {
    url += `${url.includes('?') ? '&' : '?'}teamId=${VERCEL_TEAM_ID}`;
  }

  if (opts.teamId) {
    const parsedUrl = new URL(url);
    parsedUrl.searchParams.set('teamId', opts.teamId);
    url = parsedUrl.toString();
    delete opts.teamId;
  }

  const userAgent = opts.userAgent || `client-v${pkgVersion}`;
  delete opts.userAgent;

  opts.headers = {
    ...opts.headers,
    authorization: `Bearer ${token}`,
    accept: 'application/json',
    'user-agent': userAgent,
  };

  debug(`${opts.method || 'GET'} ${url}`);
  time = Date.now();
  const res = await nodeFetch(url, opts);
  debug(`DONE in ${Date.now() - time}ms: ${opts.method || 'GET'} ${url}`);
  semaphore.release();

  return res;
};

export interface PreparedFile {
  file: string;
  sha?: string;
  size?: number;
  mode: number;
}

const isWin = process.platform.includes('win');

export const prepareFiles = (
  files: FilesMap,
  clientOptions: VercelClientOptions
): PreparedFile[] => {
  const preparedFiles: PreparedFile[] = [];
  for (const [sha, file] of files) {
    for (const name of file.names) {
      let fileName: string;

      if (clientOptions.isDirectory) {
        // Directory
        fileName =
          typeof clientOptions.path === 'string'
            ? relative(clientOptions.path, name)
            : name;
      } else {
        // Array of files or single file
        const segments = name.split(sep);
        fileName = segments[segments.length - 1];
      }

      preparedFiles.push({
        file: isWin ? fileName.replace(/\\/g, '/') : fileName,
        size: file.data?.byteLength || file.data?.length,
        mode: file.mode,
        sha: sha || undefined,
      });
    }
  }

  return preparedFiles;
};

export function createDebug(debug?: boolean) {
  if (debug) {
    return (...logs: string[]) => {
      process.stderr.write(
        [`[client-debug] ${new Date().toISOString()}`, ...logs].join(' ') + '\n'
      );
    };
  }

  return () => {};
}
type Debug = ReturnType<typeof createDebug>;
