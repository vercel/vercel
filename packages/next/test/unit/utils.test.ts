import path from 'path';
import os from 'os';
import {
  excludeFiles,
  validateEntrypoint,
  normalizePackageJson,
  getImagesConfig,
  getNextConfig,
  getServerlessPages,
  normalizePrefetches,
} from '../../src/utils';
import { FileFsRef, FileRef } from '@vercel/build-utils';
import { genDir } from '../utils';

describe('getNextConfig', () => {
  const workPath = path.join(__dirname, 'fixtures', '00-config');
  const entryPath = path.join(workPath, 'entry');

  it('should find entry file', async () => {
    const file = await getNextConfig(workPath, entryPath);
    expect(file).toMatchSnapshot();
  });

  it('should find work file second', async () => {
    const file = await getNextConfig(workPath, '/');
    expect(file).toMatchSnapshot();
  });

  it('return null on nothing', async () => {
    const file = await getNextConfig('/', '/');
    expect(file).toMatchSnapshot();
  });
});

describe('getImagesConfig', () => {
  it('should return undefined when undefined config', async () => {
    const result = await getImagesConfig(undefined);
    expect(result).toBeUndefined();
  });

  it('should return undefined when null config', async () => {
    const result = await getImagesConfig(null);
    expect(result).toBeUndefined();
  });

  it('should return undefined when empty object config', async () => {
    const result = await getImagesConfig({ images: {} });
    expect(result).toBeUndefined();
  });

  it('should return pass-through props when loader is default and unoptimized undefined', async () => {
    const images = {
      loader: 'default',
      domains: ['example.com'],
      sizes: [512, 1024],
      qualities: [25, 50, 75],
      remotePatterns: undefined,
      localPatterns: [{ search: '' }],
      formats: ['image/webp'],
      minimumCacheTTL: 60,
      dangerouslyAllowSVG: false,
      contentSecurityPolicy: undefined,
      contentDispositionType: undefined,
    };
    const result = await getImagesConfig({ images });
    expect(result).toEqual({
      domains: ['example.com'],
      sizes: [512, 1024],
      qualities: [25, 50, 75],
      remotePatterns: undefined,
      localPatterns: [{ search: '' }],
      formats: ['image/webp'],
      minimumCacheTTL: 60,
      dangerouslyAllowSVG: false,
      contentSecurityPolicy: undefined,
      contentDispositionType: undefined,
    });
  });

  it('should return pass-through props when loader is default and unoptimized false', async () => {
    const images = {
      unoptimized: false,
      loader: 'default',
      domains: ['example.com'],
      sizes: [512, 1024],
      remotePatterns: undefined,
      localPatterns: [{ pathname: '^/assets/img.png$', search: '' }],
      formats: ['image/webp'],
      minimumCacheTTL: 60,
      dangerouslyAllowSVG: false,
      contentSecurityPolicy: undefined,
      contentDispositionType: 'attachment',
    };
    const result = await getImagesConfig({ images });
    expect(result).toEqual({
      domains: ['example.com'],
      sizes: [512, 1024],
      remotePatterns: undefined,
      localPatterns: [{ pathname: '^/assets/img.png$', search: '' }],
      formats: ['image/webp'],
      minimumCacheTTL: 60,
      dangerouslyAllowSVG: false,
      contentSecurityPolicy: undefined,
      contentDispositionType: 'attachment',
    });
  });

  it('return return undefined when loader is default and unoptimized true', async () => {
    const images = {
      unoptimized: true,
      loader: 'default',
      domains: ['example.com'],
      sizes: [512, 1024],
      qualities: [70, 80, 90],
      remotePatterns: undefined,
      localPatterns: [{ search: '' }],
      formats: ['image/webp'],
      minimumCacheTTL: 60,
      dangerouslyAllowSVG: false,
      contentSecurityPolicy: undefined,
      contentDispositionType: undefined,
    };
    const result = await getImagesConfig({ images });
    expect(result).toBeUndefined();
  });
});

describe('excludeFiles', () => {
  it('should exclude files', () => {
    const files = {
      'pages/index.js': new FileRef({ digest: 'index' }),
      'package.json': new FileRef({ digest: 'package' }),
      'package-lock.json': new FileRef({ digest: 'package-lock' }),
    };
    const result = excludeFiles(
      files,
      filePath => filePath === 'package-lock.json'
    );
    expect(result['pages/index.js']).toBeDefined();
    expect(result['package.json']).toBeDefined();
    expect(result['package-lock.json']).toBeUndefined();
  });
});

describe('validateEntrypoint', () => {
  it('should allow package.json', () => {
    expect(validateEntrypoint('package.json')).toBeUndefined();
  });
  it('should allow nested package.json', () => {
    expect(validateEntrypoint('frontend/package.json')).toBeUndefined();
  });
  it('should allow next.config.js', () => {
    expect(validateEntrypoint('next.config.js')).toBeUndefined();
  });
  it('should allow nested next.config.js', () => {
    expect(validateEntrypoint('frontend/next.config.js')).toBeUndefined();
  });
  it('should not allow pages/index.js', () => {
    expect(() => validateEntrypoint('pages/index.js')).toThrow();
  });
});

describe('normalizePackageJson', () => {
  it('should work without a package.json being supplied', () => {
    const result = normalizePackageJson();
    expect(result).toEqual({
      dependencies: {
        'next-server': 'v7.0.2-canary.49',
        react: 'latest',
        'react-dom': 'latest',
      },
      devDependencies: {
        next: 'v7.0.2-canary.49',
      },
      scripts: {
        'now-build':
          'NODE_OPTIONS=--max_old_space_size=3000 next build --lambdas',
      },
    });
  });

  it('should work with a package.json being supplied', () => {
    const defaultPackage = {
      dependencies: {
        'next-server': 'v7.0.2-canary.49',
        react: 'latest',
        'react-dom': 'latest',
      },
      devDependencies: {
        next: 'v7.0.2-canary.49',
      },
      scripts: {
        'now-build': 'next build',
      },
    };
    const result = normalizePackageJson(defaultPackage);
    expect(result).toEqual({
      dependencies: {
        'next-server': 'v7.0.2-canary.49',
        react: 'latest',
        'react-dom': 'latest',
      },
      devDependencies: {
        next: 'v7.0.2-canary.49',
      },
      scripts: {
        'now-build':
          'NODE_OPTIONS=--max_old_space_size=3000 next build --lambdas',
      },
    });
  });

  it('should force next@canary to be a devDependency', () => {
    const defaultPackage = {
      dependencies: {
        react: 'latest',
        'react-dom': 'latest',
        next: 'latest',
      },
    };
    const result = normalizePackageJson(defaultPackage);
    expect(result).toEqual({
      dependencies: {
        'next-server': 'v7.0.2-canary.49',
        react: 'latest',
        'react-dom': 'latest',
      },
      devDependencies: {
        next: 'v7.0.2-canary.49',
      },
      scripts: {
        'now-build':
          'NODE_OPTIONS=--max_old_space_size=3000 next build --lambdas',
      },
    });
  });

  it('should force next-server@canary to be a dependency', () => {
    const defaultPackage = {
      dependencies: {
        react: 'latest',
        'react-dom': 'latest',
        next: 'latest',
      },
    };
    const result = normalizePackageJson(defaultPackage);
    expect(result).toEqual({
      dependencies: {
        'next-server': 'v7.0.2-canary.49',
        react: 'latest',
        'react-dom': 'latest',
      },
      devDependencies: {
        next: 'v7.0.2-canary.49',
      },
      scripts: {
        'now-build':
          'NODE_OPTIONS=--max_old_space_size=3000 next build --lambdas',
      },
    });
  });

  it('should force now-build script', () => {
    const defaultPackage = {
      dependencies: {
        react: 'latest',
        'react-dom': 'latest',
        next: 'latest',
      },
    };
    const result = normalizePackageJson(defaultPackage);
    expect(result).toEqual({
      dependencies: {
        'next-server': 'v7.0.2-canary.49',
        react: 'latest',
        'react-dom': 'latest',
      },
      devDependencies: {
        next: 'v7.0.2-canary.49',
      },
      scripts: {
        'now-build':
          'NODE_OPTIONS=--max_old_space_size=3000 next build --lambdas',
      },
    });
  });

  // https://github.com/vercel/next.js/issues/5700
  it('should normalize user report vercel/next.js#5700 correctly', () => {
    const defaultPackage = {
      version: '1.0.0',
      scripts: {
        dev: 'next',
        build: 'next build',
        start: 'next start',
        test: "xo && stylelint './pages/**/*.js' && jest",
      },
      main: 'pages/index.js',
      license: 'MIT',
      devDependencies: {
        'babel-plugin-styled-components': '^1.8.0',
        'eslint-config-xo-react': '^0.17.0',
        'eslint-plugin-react': '^7.11.1',
        jest: '^23.6.0',
        'jest-styled-components': '^6.3.1',
        'react-test-renderer': '^16.6.3',
        stylelint: '^9.8.0',
        'stylelint-config-recommended': '^2.1.0',
        'stylelint-config-styled-components': '^0.1.1',
        'stylelint-processor-styled-components': '^1.5.1',
        xo: '^0.23.0',
      },
      dependencies: {
        consola: '^2.2.6',
        fontfaceobserver: '^2.0.13',
        next: '^7.0.2',
        react: '^16.6.3',
        'react-dom': '^16.6.3',
        'styled-components': '^4.1.1',
      },
      xo: {
        extends: 'xo-react',
        envs: 'browser',
        esnext: true,
        ignores: [
          'test',
          'pages/_document.js',
          'pages/index.js',
          'pages/home.js',
        ],
        rules: {
          'react/no-unescaped-entities': null,
        },
      },
      jest: {
        testEnvironment: 'node',
      },
    };
    const result = normalizePackageJson(defaultPackage);
    expect(result).toEqual({
      version: '1.0.0',
      scripts: {
        dev: 'next',
        build: 'next build',
        'now-build':
          'NODE_OPTIONS=--max_old_space_size=3000 next build --lambdas',
        start: 'next start',
        test: "xo && stylelint './pages/**/*.js' && jest",
      },
      main: 'pages/index.js',
      license: 'MIT',
      devDependencies: {
        'babel-plugin-styled-components': '^1.8.0',
        'eslint-config-xo-react': '^0.17.0',
        'eslint-plugin-react': '^7.11.1',
        jest: '^23.6.0',
        'jest-styled-components': '^6.3.1',
        'react-test-renderer': '^16.6.3',
        stylelint: '^9.8.0',
        'stylelint-config-recommended': '^2.1.0',
        'stylelint-config-styled-components': '^0.1.1',
        'stylelint-processor-styled-components': '^1.5.1',
        next: 'v7.0.2-canary.49',
        'next-server': undefined,
        xo: '^0.23.0',
        consola: '^2.2.6',
        fontfaceobserver: '^2.0.13',
        'styled-components': '^4.1.1',
      },
      dependencies: {
        'next-server': 'v7.0.2-canary.49',
        react: '^16.6.3',
        'react-dom': '^16.6.3',
      },
      xo: {
        extends: 'xo-react',
        envs: 'browser',
        esnext: true,
        ignores: [
          'test',
          'pages/_document.js',
          'pages/index.js',
          'pages/home.js',
        ],
        rules: {
          'react/no-unescaped-entities': null,
        },
      },
      jest: {
        testEnvironment: 'node',
      },
    });
  });
});

describe('getServerlessPages', () => {
  it('should gather all pages correctly', async () => {
    const dir = await genDir({
      '.next/server/pages/_app.js': 'test',
      '.next/server/pages/_error.js': 'test',
      '.next/server/app/page.js': 'test',
      '.next/server/app/favicon.ico/route.js': 'test',
    });

    const { pages, appPaths } = await getServerlessPages({
      pagesDir: path.resolve(path.join(dir, '.next/server/pages')),
      entryPath: os.tmpdir(),
      outputDirectory: os.tmpdir(),
      appPathRoutesManifest: {
        '/_not-found': '/_not-found',
        '/favicon.ico/route': '/favicon.ico',
        '/page': '/',
      },
    });

    expect(Object.keys(pages)).toEqual(['_app.js', '_error.js']);
    expect(Object.keys(appPaths)).toEqual(['favicon.ico.js', 'index.js']);
  });
});

describe('normalizePrefetches', () => {
  it('should properly prefix prefetches with `__`', async () => {
    const dummyFile = new FileFsRef({ fsPath: __dirname });

    const appRscPrefetches = {
      'index.prefetch.rsc': dummyFile,
      'index/index.prefetch.rsc': dummyFile,
      'foo.prefetch.rsc': dummyFile,
      'foo/index.prefetch.rsc': dummyFile,
      'foo/bar/baz.prefetch.rsc': dummyFile,
    };

    const updatedPrefetches = normalizePrefetches(appRscPrefetches);

    expect(Object.keys(updatedPrefetches)).toEqual([
      '__index.prefetch.rsc',
      'index/index.prefetch.rsc',
      'foo.prefetch.rsc',
      'foo/index.prefetch.rsc',
      'foo/bar/baz.prefetch.rsc',
    ]);
  });
});
