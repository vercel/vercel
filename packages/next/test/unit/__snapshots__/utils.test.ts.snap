// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNextConfig return null on nothing 1`] = `null`;

exports[`getNextConfig should find entry file 1`] = `
"module.exports = {};
"
`;

exports[`getNextConfig should find work file second 1`] = `
"module.exports = { target: 'serverless' };
"
`;
